### DISCUSSION TITLE
How to properly patch packages?

### DISCUSSION THREAD

**Discussion by swazrgb** [Morganamilo/paru#835] [Q&A]
During package review you can modify the PKGBUILD and commit the result. Doing so works fine, until the PKGBUILD changes upstream. Then the following happens:

```
error: could not apply 7ad8538... custom patch
    hint: Resolve all conflicts manually, mark them as resolved with
    hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
    hint: You can instead skip this commit: run "git rebase --skip".
    hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
    Could not apply 7ad8538... custom patch
```

After resolving the conflict in `/home/user/.cache/paru/clone/qemu-git` paru then starts erroring with:

```
error: command failed: /home/user/.cache/paru/clone/qemu-git: git -c user.email=aur -c user.name=aur merge --no-edit --no-ff --no-commit:
```

Am I not supposed to change PKGBUILD during review? I thought this was the purpose of `--fm`. How should I be patching packages?

So far the only way I managed to get out of this mess is to delete `/home/user/.cache/paru/clone/qemu-git` and start over, reapplying my patch. Which is obviously a pain.

**Comment by Morganamilo** [comment]
Paru will auto rebase as long as there's not a conflict. Clearly there was one here. If you solve the conflicts it should work

**Reply by swazrgb** [reply]
Yes, resolving the conflict produces the above merge error. These are my steps:

1. Install package using paru (in my case qemu-git)
2. During review phase modify the PKGBUILD and commit the result
3. Install the package. Everything is great.
4. When a conflicting change happens in the upstream PKGBUILD paru initially errors with a rebase error. This is completely expected.
5. I resolve the conflict, git add the file, and do `git rebase --continue`
6. I re-run paru. It now fails with the mentioned merge error.
7. I resolve the merge conflict. I have both tried committing the merge and just leaving it `git add`ed
8. I re-run paru. It fails again either with the rebase or merge error, I'm unsure which, sorry.
9. I've tried repeating resolving the merge, even installing git rerere to auto-resolve it, but no matter what, it keeps failing on a conflict

So I'm not sure what I'm doing wrong there. I know how to use git, am resolving the conflict correctly, but some of the git commands that paru uses seem to be discarding my resolution.

**Reply by max-kamps** [reply]
I'm running into the same issue today. I manually resolved the merge conflict. So `git status` returns `Your branch is ahead of 'origin/master' by 1 commit. nothing to commit, working tree clean`, and `git pull` returns `Already up to date.`

But then when I re-run `paru`, I get
```
PKGBUILDs up to date
error: command failed: /home/max/.cache/paru/clone/vimix-icon-theme: git -c user.email=aur -c user.name=aur merge --no-edit --no-ff --no-commit:

```
(yes, it ends with an empty line and no error message)
And then my merge resolution has been discarded.

**Comment by HaleTom** [comment]
Use PreBuildCommand hook to call customizepkg: https://github.com/Morganamilo/paru/pull/539

https://github.com/ava1ar/customizepkg

HOWTO: https://www.supertechcrew.com/customizing-aur-arch-package-with-customizepkg/
